### PR TITLE
CU-25ye81h CU-1yyr5hv CU-1yyr5fv xcmp kusd picasso

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
Signed-off-by: Dzmitry Lahoda <dzmitry@lahoda.pro>

- hardened ED - made mutli ED non zero
- allowed to transfer from Karura
- reused XCMP in Picasso

### CODE REVIEW

File: integration-tests/local-integration-tests/awesome-xcmp.md [hunk 1]
```diff
@@ -1,8 +1,10 @@
-## Readings
+## Overview
+
+Resources which allows to grapsh XCM. General undrestanding how bridges work will help too.
```
Comment (vivekvpandya) [inline integration-tests/local-integration-tests/awesome-xcmp.md:3]: typo: `grapsh` -> `grasp` .
suggestion: reword the sentence `Resources which helps understanding XCM. `

File: integration-tests/local-integration-tests/src/kusama_test_net.rs [hunk 2]
```diff
@@ -1,7 +1,17 @@
 //! Setup of Picasso running as if it is on Kusama relay
 use common::AccountId;
 use cumulus_primitives_core::ParaId;
-use dali_runtime as picasso_runtime;
+
+#[cfg(feature = "dali")]
+use dali_runtime as sibling_runtime;
+#[cfg(feature = "dali")]
```
Comment (vivekvpandya) [inline integration-tests/local-integration-tests/src/kusama_test_net.rs:7]: is this correct? shouldn't one of above check for feature != dali?
if it is correct could you please explain more details? @dzmitry-lahoda

Comment (dzmitry-lahoda) [reply integration-tests/local-integration-tests/src/kusama_test_net.rs:7]: there will be composable in future.

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1yyr5hv xcmp: make multi existential balance and fixe ED tests parts](https://app.clickup.com/t/1yyr5hv)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1yyr5fv xcmp: use commong runtime parts to setup runtime in picasso](https://app.clickup.com/t/1yyr5fv)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-25ye81h xcmp: karura: make aUSD transferable in simulator](https://app.clickup.com/t/25ye81h)

**Comment by hussein-aitlahcen** [comment]
@dzmitry-lahoda Can we merge main back into this or rebase so that we generate new weights :)?

**Review by hussein-aitlahcen — APPROVED** [review]
LGTM, quick question: do we currently handle decimal normalization somewhere?

File: runtime/common/src/xcmp/mod.rs [hunk 3]
```diff
@@ -1,6 +1,349 @@
...
+use xcm_builder::*;
+use xcm_executor::{
+	traits::{FilterAssetLocation, ShouldExecute, WeightTrader},
+	*,
+};
 
-use frame_support::{dispatch::Weight, parameter_types};
 parameter_types! {
 	pub const BaseXcmWeight: Weight = 100_000_000;
 }
```
Comment (KaiserKarel) [inline runtime/common/src/xcmp/mod.rs:29]: Can we add a comment here describing why we chose this weight?

Comment (dzmitry-lahoda) [reply runtime/common/src/xcmp/mod.rs:29]: will add. in short, this is taken from acala/hydra.

File: runtime/common/src/lib.rs [hunk 4]
```diff
@@ -154,19 +148,42 @@ impl MinimalOracle for PriceConverter {
...
+#[cfg(feature = "runtime-benchmarks")]
+pub fn multi_existential_deposits(_currency_id: &CurrencyId) -> Balance {
+	// ISSUE:
+	// 1. runing only collective pallet does not lead to issue (we cannot control other pallets
+	// always) 2. pallet benchmark does not contain direct manipulation of deposits
+	// 3. increatng caller balance does not help to fix issue
+	// 4. so it contains weight return to take from caller
+	// 2022-03-14 20:50:19 Running Benchmark: collective.set_members 2/1 1/1
+	// Error:
+	//   0: Invalid input: Account cannot exist with the funds that would be given
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply runtime/common/src/lib.rs:176]: we use 3rd party pallets. when we set non zero non native deposits, benching them fail. one specific pallet. collective.  i looked into it code and do not know why it fails so.

[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply runtime/common/src/lib.rs:176]: so, some pallets we used assume that ed is zero in benches

File: runtime/common/src/xcmp/mod.rs [hunk 5]
```diff
@@ -1,6 +1,349 @@
...
+
+impl ShouldExecute for XcmpDebug {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		log::trace!(target: "xcmp::should_execute", "{:?} {:?} {:?} {:?}", origin, message, max_weight, weight_credit);
+		Err(())
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply runtime/common/src/xcmp/mod.rs:44]: err is false here, join3d with other filters by or.

[the remark this answers is not in the document]
Comment (hussein-aitlahcen) [reply runtime/common/src/xcmp/mod.rs:44]: Alright makes sense.

**Comment by dzmitry-lahoda** [comment]
@hussein-aitlahcen assets we transfer are 12 decimals now. I have not see explicit hooks to handle decimals not saw it in acala. 
we added decimals into foreign assets pallet, so we will abe able to handle that in orcle. but for others assets it pose the problem. we may end up need to hook into executor to check each instruction to map decimals. and Transact instruction is binary block. so what we end up- the send must transform decimals into 12 before sending to our chain. would they agree?

File: runtime/common/src/xcmp/mod.rs (hunk 3 above)
Comment (dzmitry-lahoda) [reply runtime/common/src/xcmp/mod.rs:29]: weight of messages it self is added on top. it also kind of arbitrary. also once I saw mention of weights for XCMP, but never seen any benches in any chain

**Comment by dzmitry-lahoda** [comment]
@KaiserKarel current weights likely are too high overall.
examples

fn transfer() -> Weight {
		(101_289_000 as Weight)

we hafe 100_000_000 for empty XCMP + extra on top for each insutction 


	// One XCM operation is 200_000_000 weight, cross-chain transfer ~= 2x of transfer.
	pub const UnitWeightCost: Weight = 200_000_000;

so at least we will take 300_000_000 for any 1 instruction at least.

It seems kind of OK. But I think that value should be some how calculated later (they have ideas of shallow + deep weights in XCMP)

File: runtime/picasso/build.rs [hunk 6]
```diff
@@ -1,9 +1,10 @@
...
-		.import_memory()
-		.build()
+	#[cfg(feature = "wasm-builder")]
+	{
+		substrate_wasm_builder::WasmBuilder::new()
+			.with_current_project()
+			.export_heap_base()
+			.import_memory()
+			.build()
+	}
```
Comment (hussein-aitlahcen) [inline runtime/picasso/build.rs:9]: Rationale behind adding this feature? It's not harmful but I am curious :)

**Review by hussein-aitlahcen — DISMISSED** [review]

File: runtime/picasso/build.rs (hunk 6 above)
Comment (dzmitry-lahoda) [reply runtime/picasso/build.rs:9]: to avoid building wasm when running native

Comment (dzmitry-lahoda) [reply runtime/picasso/build.rs:9]: this speed ups build of simulator tests

Comment (hussein-aitlahcen) [reply runtime/picasso/build.rs:9]: Hm, isn't this avoided by SKIP_WASM_BUILD?

**Review by Dom-Roth — APPROVED** [review]

**Comment by dzmitry-lahoda** [comment]
oh, bad. forgot to sign during rebase...

**Review by Dom-Roth — APPROVED** [review]

**Review by KaiserKarel — APPROVED** [review]

**Pull request merged**